ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,222 @@
+    this.contactedInCurrentExecution = new ArrayList<>();
+
+    commenceTimeoutTask();
+  }
+
+  private void commenceTimeoutTask() {
+    vertx.setPeriodic(
+        TIMEOUT_TASK_DELAY,
+        v -> {
+          List<OutstandingRequest> outstandingRequestListCopy =
+              new ArrayList<>(outstandingRequestList);
+
+          for (OutstandingRequest outstandingRequest : outstandingRequestListCopy) {
+            if (outstandingRequest.isExpired()) {
+              List<Peer> queryCandidates = determineFindNodeCandidates(anteList.size());
+              for (Peer candidate : queryCandidates) {
+                if (!contactedInCurrentExecution.contains(candidate.getId())
+                    && !outstandingRequestList.contains(new OutstandingRequest(candidate))) {
+                  executeFindNodeRequest(candidate);
+                }
+              }
+              outstandingRequestList.remove(outstandingRequest);
+            }
+          }
+        });
+  }
+
+  private void executeFindNodeRequest(final Peer peer) {
+    BytesValue peerId = peer.getId();
+    outstandingRequestList.add(new OutstandingRequest(peer));
+    contactedInCurrentExecution.add(peerId);
+    neighborFinder.issueFindNodeRequest(peer);
+  }
+
+  /**
+   * The lookup initiator starts by picking CONCURRENT_REQUEST_LIMIT closest nodes to the target it
+   * knows of. The initiator then issues concurrent FindNode packets to those nodes.
+   */
+  private void initiatePeerRefreshCycle(final List<Peer> peers) {
+    for (Peer peer : peers) {
+      if (!contactedInCurrentExecution.contains(peer.getId())) {
+        executeFindNodeRequest(peer);
+      }
+      // The lookup terminates when the initiator has queried
+      // and gotten responses from the k closest nodes it has seen.
+    }
+  }
+
+  void digestNeighboursPacket(final NeighborsPacketData neighboursPacket, final Peer peer) {
+    if (outstandingRequestList.contains(new OutstandingRequest(peer))) {
+      List<Peer> receivedPeerList = neighboursPacket.getNodes();
+      for (Peer receivedPeer : receivedPeerList) {
+        if (!peerBlacklist.contains(receivedPeer)) {
+          bondingAgent.performBonding(receivedPeer);
+          anteList.add(new PeerDistance(receivedPeer, distance(target, receivedPeer.getId())));
+        }
+      }
+      outstandingRequestList.remove(new OutstandingRequest(peer));
+      performIteration();
+    }
+  }
+
+  private List<Peer> determineFindNodeCandidates(final int threshold) {
+    anteList.sort(
+        (peer1, peer2) -> {
+          if (peer1.getDistance() > peer2.getDistance()) return 1;
+          if (peer1.getDistance() < peer2.getDistance()) return -1;
+          return 0;
+        });
+    return anteList.subList(0, threshold).stream().map(PeerDistance::getPeer).collect(toList());
+  }
+
+  private void performIteration() {
+    if (outstandingRequestList.isEmpty()) {
+      List<Peer> queryCandidates = determineFindNodeCandidates(CONCURRENT_REQUEST_LIMIT);
```

Comment:
If we query 3 other peers for each response we receive, the number of concurrent requests will be 3 in the first round, 9 in the second and increase exponentially.  Should we be querying the 3 closest from each response or the 3 closest from all the responses from that round?

The current behaviour may be ok - having a lot of outstanding requests isn't that big a deal since we don't have to do anything while we wait for them but if do increase exponentially we need a fairly low limit on the number of rounds we do so we don't flood the network.

Reply:
As it's currently configured I'm pretty sure it queries the 3 closest peers from a round. If we take (**A**) as a bootstrap node, and (**A**) sends us back a packet that tells us about (**B**), (**C**), (**D**) and (**E**), we'll consider these peers and deposit them all into the *anteList*. As we had one *OutstandingRequest*, to (**A**), which we've received a response from, our *OutstandingRequestList* will be empty, so we'll start another round. Let's say that our closest peers are (**B**), (**C**) and (**D**), we'll issue requests and (assuming non of them expire), we'll received responses from each one of the, each of which with *n* peers, all of which will be deposited into the  *anteList*. When we've gotten responses from all of them, that is when our *OutstandingRequestList* will again be empty, which is what will precipitate a fresh round of 3 requests.